spec/lib/mailchimp_list_synchronizer_spec.rb

```diff
@@ -77,91 +56,227 @@
       ENV["SETTINGS__MAILCHIMP__API_KEY"] = "KEY"
     end
 
-    it "subscribes users to MailChimp who appear in the database, but not in the mailing list" do
-      expect(mailchimp_client_lists).to receive(:set_list_member).with(
-        "list-1",
-        anything,
-        {
-          "email_address" => "subscribe@domain.org",
-          "status" => "subscribed",
-        },
-      )
+    RSpec.shared_examples "it subscribes the user" do
+      it "subscribes the user" do
+        expect(mailchimp_client_lists).to receive(:set_list_member).with(
+          "list-1",
+          anything,
+          {
+            "email_address" => "user@domain.org",
+            "status" => "subscribed",
+          },
+        )
 
-      described_class.synchronize(list_id: "list-1", users_to_synchronize:)
+        described_class.synchronize(list_id: "list-1", users_to_synchronize:)
+      end
     end
 
-    it "does not subscribe users who are in the database but lack access to GOV.UK Forms" do
-      expect(mailchimp_client_lists).not_to receive(:set_list_member).with(
-        "list-1",
-        anything,
-        {
-          "email_address" => "user.without.access@domain.org",
-          "status" => "subscribed",
-        },
-      )
+    RSpec.shared_examples "it archives the user" do
+      it "archives the user" do
+        archived_email_hash = Digest::MD5.hexdigest "user@domain.org"
 
-      described_class.synchronize(list_id: "list-1", users_to_synchronize:)
+        expect(mailchimp_client_lists).to receive(:delete_list_member).with("list-1", archived_email_hash)
+
+        described_class.synchronize(list_id: "list-1", users_to_synchronize:)
+      end
     end
 
-    it "archives users from MailChimp who do not appear in the database, but do appear in the mailing list" do
-      archived_email_hash = Digest::MD5.hexdigest "archive@domain.org"
+    RSpec.shared_examples "it does not subscribe or archive the user" do
+      it "does not subscribe or archive the user" do
+        archived_email_hash = Digest::MD5.hexdigest "user@domain.org"
+
+        expect(mailchimp_client_lists).not_to receive(:set_list_member).with(
```

Comment:
I wonder whether this expectation is quite brittle. Because we're asserting on the method not receiving a call with very specific arguments, this might continue to pass if the logic under test doesn't work as expected but the hash we pass to the method changes slightly.

Could we instead either assert that `:set_list_member` is either not called with any arguments, or if it is called do an expectation on the arguments that they don't match the ones we don't want it to be called with.

Something like
```
expect(mailchimp_client_lists).to receive(:set_list_member) do |list_id, subscriber_hash, body|
  expect(body[:email_address]).not_to eq("user@domain.org")
end
```

I'm not convinced what I've suggested is great, there might be a better way to do it.

Reply:
Actually it's probably better if it is called to check it is called exactly once and with the expected arguments. Something like this?

```
expect(mailchimp_client_lists).to receive(:set_list_member).once do |list_id, subscriber_hash, body|
  expect(body["email_address"]).to eq("some_other_user@domain.org")
end
```

Reply:
Yeah, I think this is valid and that seems like a good approach - thanks for the suggestion!

Reply:
Have updated!